File: python/problem1376.py
class Solution:
    def numOfMinutes(self, n: int, headID: int, manager: List[int], informTime: List[int]) -> int:
        toRootMins = [0 for _ in range(n)]
        def timeToRoot(id):
            if id == headID:            #总负责人
                return 0
            if toRootMins[id] != 0:     #已知当前员工节点到总负责人节点时间，直接返回
                return toRootMins[id]
            #已知当前员工节点的上司节点到总负责人节点时间，那就在这基础上加上上司节点到当前员工节点的时间，否则需要递归计算
            toRootMins[id] = (toRootMins[manager[id]] + informTime[manager[id]]) if toRootMins[manager[id]] != 0 else (timeToRoot(manager[id]) + informTime[manager[id]])
            return toRootMins[id]
        maxMins = 0
        for i in range(n):
            if informTime[i] == 0:      #只需要考虑叶子节点也即最下层员工节点
                tmp = timeToRoot(i)
                if tmp > maxMins:
                    maxMins = tmp
        return maxMins

File: Java/problem1376.java
class Solution {
    private Map<Integer, List<Integer>> manager2sub;

    private int maxMins(int id, int[] informTime) {
        if (informTime[id] > 0) {
            List<Integer> subs = manager2sub.get(id);
            int maxTime = 0;
            for (int i : subs) {
                int subTime = maxMins(i, informTime);
                if (subTime > maxTime) {
                    maxTime = subTime;
                }
            }
            return informTime[id] + maxTime;
        }
        else {
            return 0;
        }
    }
    
    public int numOfMinutes(int n, int headID, int[] manager, int[] informTime) {
        manager2sub = new HashMap<>();
        for (int i = 0; i < n; ++i) {
            List<Integer> tmp = manager2sub.containsKey(manager[i]) ? manager2sub.get(manager[i]) : new LinkedList<>();
            tmp.add(i);
            manager2sub.put(manager[i], tmp);
        }
        return maxMins(headID, informTime);
    }
}
